Adding example of fold for sink

muskan3006 opened the pull request:
<!--
  Are there any relevant issues / PRs / mailing lists discussions?
  Please reference them here - but don't use `fixes` notation.
-->
References #25468

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

raboof commented:
Thanks for having a look at this! You can auto-format the docs with `sbt akka-docs/test:scalafmt`.

raboof commented:
Oh, and for the Java code it's `sbt akka-docs/test:javafmt`

raboof commented on akka-docs/src/test/java/jdocs/stream/operators/SinkDocExamples.java:
```diff
@@ -6,122 +6,126 @@
 import akka.actor.ActorSystem;
-
-import akka.stream.javadsl.Flow;
+import akka.japi.Pair;
```
It would be better not to rearrange the imports, since now the 'Pair' import is no longer part of the `#takeLast-operator-example` snippet.

muskan3006 replied:
Okay I'll resolve this

muskan3006 commented:
sbt is throwing some error in this repo, I am trying to resolve it

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

muskan3006 commented:
When I run sbt akka-docs/test:scalafmt this command, there are around 200 files that are reformatted, should I include all of them in my PR?

raboof commented:
> When I run sbt akka-docs/test:scalafmt this command, there are around 200 files that are reformatted, should I include all of them in my PR?

Hmm, no, that doesn't sound right... sorry about the hassle! I can run the formatter for you and amend this PR, if you like?

It would be good to understand why it's reformatting so much - but I'm not sure where to start...

muskan3006 commented:
> > When I run sbt akka-docs/test:scalafmt this command, there are around 200 files that are reformatted, should I include all of them in my PR?
> 
> Hmm, no, that doesn't sound right... sorry about the hassle! I can run the formatter for you and amend this PR, if you like?
> 
> It would be good to understand why it's reformatting so much - but I'm not sure where to start...

Yes, please do that and thanks.

muskan3006 commented:
Also in sbtops we have some options but they are unrecognizible atleast in my system(Windows 10). When I hit sbt it shows this error. For resolving this I had to remove -J from the sbtops and then it worked.

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

raboof commented:
> Yes, please do that and thanks.

No problem! It was almost there, I just added a file header.

> Also in sbtops we have some options but they are unrecognizible atleast in my system(Windows 10).

Hmm, that seems to be a known problem in sbt: https://github.com/sbt/sbt/issues/5179

> At least one pull request committer is not linked to a user

It looks like you used muskan.gupta3006@google.com in the commits, but didn't add that account to your github profile yet. Could you add it as described at https://docs.github.com/en/github/setting-up-and-managing-your-github-user-account/adding-an-email-address-to-your-github-account ?

muskan3006 commented:
> > Yes, please do that and thanks.
> 
> No problem! It was almost there, I just added a file header.
> 
> > Also in sbtops we have some options but they are unrecognizible atleast in my system(Windows 10).
> 
> Hmm, that seems to be a known problem in sbt: [sbt/sbt#5179](https://github.com/sbt/sbt/issues/5179)
> 
> > At least one pull request committer is not linked to a user
> 
> It looks like you used [muskan.gupta3006@google.com](mailto:muskan.gupta3006@google.com) in the commits, but didn't add that account to your github profile yet. Could you add it as described at https://docs.github.com/en/github/setting-up-and-managing-your-github-user-accoun

t/adding-an-email-address-to-your-github-account ?

> > Yes, please do that and thanks.
> 
> No problem! It was almost there, I just added a file header.
> 
> > Also in sbtops we have some options but they are unrecognizible atleast in my system(Windows 10).
> 
> Hmm, that seems to be a known problem in sbt: [sbt/sbt#5179](https://github.com/sbt/sbt/issues/5179)
> 
> > At least one pull request committer is not linked to a user
> 
> It looks like you used [muskan.gupta3006@google.com](mailto:muskan.gupta3006@google.com) in the commits, but didn't add that account to your github profile yet. Could you add it as described at https://docs.github.com/en/github/setting-up-and-managing-your-github-user-account/adding-an-email-address-to-your-github-account ?

the account is muskan.gupta3006@gmail.com and it is added to my github

raboof commented:
> the account is muskan.gupta3006@gmail.com and it is added to my github

Ah, but in the commits it is `@google.com`:

```
commit ed1da2ac0bccd98adeaef189e452a92a100faabe
Author: Muskan Gupta <muskan.gupta3006@google.com>
Date:   Wed Sep 16 16:16:30 2020 +0530

    Adding example of fold for sink
```

It seems your `git config user.email` is set incorrectly. You can fix it with `git config --global user.email "MY_NAME@example.com"`, but you probably also want to fix the commits. You can do that with 'rebase' and then force-push to the branch.

muskan3006 commented:
> > the account is [muskan.gupta3006@gmail.com](mailto:muskan.gupta3006@gmail.com) and it is added to my github
> 
> Ah, but in the commits it is `@google.com`:
> 
> ```
> commit ed1da2ac0bccd98adeaef189e452a92a100faabe
> Author: Muskan Gupta <muskan.gupta3006@google.com>
> Date:   Wed Sep 16 16:16:30 2020 +0530
> 
>     Adding example of fold for sink
> ```
> 
> It seems your `git config user.email` is set incorrectly. You can fix it with `git config --global user.email "MY_NAME@example.com"`, but you probably also want to fix the commits. You can do that with 'rebase' and then force-push to the branch.

Done

raboof commented:
> Done

It's probably configured correctly for new commits, but the commits in this branch still have the wrong address. The instructions at https://stackoverflow.com/questions/3042437/how-to-change-the-commit-author-for-one-specific-commit seem relevant here.

raboof commented:
Sorry about all the administrative hurdles, now that this is set up correctly it will probably be much smoother next time :smile:

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

muskan3006 commented:
> Sorry about all the administrative hurdles, now that this is set up correctly it will probably be much smoother next time 😄

Yes, thanks for this much guidance otherwise it would have been really difficult for me

muskan3006 committed:
Adding example of fold for sink

muskan3006 committed:
Formatted Flow.scala

muskan3006 committed:
Formatted Fold.scala

muskan3006 committed:
Formatting files

raboof committed:
sbt akka-docs/test:headerCreate

raboof reviewed (approved):
Nice, thank you!

muskan3006 commented:
> > Yes, please do that and thanks.
> 
> No problem! It was almost there, I just added a file header.
 
Can you tell me how you resolved this? The same issue is in my other PR as well

raboof commented:
> > No problem! It was almost there, I just added a file header.
> 
> Can you tell me how you resolved this? The same issue is in my other PR as well

I added the header with `sbt headerCreate test:headerCreate`. If you mean #29623, that one has a different problem: there the index is out of sync with the operator page (https://github.com/akka/akka/pull/29623/files#r490769206)

johanandren commented on akka-docs/src/test/scala/docs/stream/operators/sink/Fold.scala:
```diff
@@ -0,0 +1,21 @@
+
+  def foldExample: Future[Int] = {
+    //#fold
+    Source(1 to 100).runWith(Sink.fold(0)(_ + _))
```
I think it'd that although the _ + _ is fine in general it would be good to give the parameters names in an example as they help understand what is what. Names should be the same as in the Java example.

johanandren commented on akka-docs/src/test/java/jdocs/stream/operators/SinkDocExamples.java:
```diff
@@ -105,6 +105,12 @@ static void lastOptionExample() {
 
+  static void foldExample() {
+    // #fold
+    Source.range(1, 100).runWith(Sink.fold(0, (res, number) -> res + number), system);
```
I'd go with `acc` (as in accumlated) or `sum` and `element` so that it is more clear what parameter is what.

johanandren commented on akka-docs/src/test/java/jdocs/stream/operators/SinkDocExamples.java:
```diff
@@ -105,6 +105,12 @@ static void lastOptionExample() {
+  static void foldExample() {
+    // #fold
+    Source.range(1, 100).runWith(Sink.fold(0, (res, number) -> res + number), system);
+    // #fold
```
I think it would be good with another few lines actually consuming the result (that would also require that you catch the return value in a variable and make the type of the materialized value explicit.

For an existing operator sample that does this, you can look at Sink.last.

johanandren commented on akka-docs/src/test/scala/docs/stream/operators/sink/Fold.scala:
```diff
@@ -0,0 +1,21 @@
+  def foldExample: Future[Int] = {
+    //#fold
+    Source(1 to 100).runWith(Sink.fold(0)(_ + _))
+    //Future(Success(5050))
```
Same as for Java, given that this will be read by users that are potentially newcomers to even async programming it is good with complete act-on-materialized-value logic.

johanandren reviewed (commented):
Added some ideas improvement!

muskan3006 committed:
Modify example

muskan3006 committed:
Merge branch 'feature/add-fold-example' of github.com:muskan3006/akka into feature/add-fold-example

muskan3006 committed:
Formatting file

johanandren reviewed (commented):
One `sbt akka-docs/test:javafmt` more and this is good to go!

muskan3006 committed:
Formatting java file

muskan3006 commented:
> One `sbt akka-docs/test:javafmt` more and this is good to go!

Done

johanandren reviewed (approved):
LGTM, thanks @muskan3006 !
